src/sys/windows/selector.rs

```diff
@@ -351,43 +400,104 @@ impl Selector {
 
 #[derive(Debug)]
 pub struct SelectorInner {
-    cp: CompletionPort,
-    active_poll_count: AtomicUsize,
-    update_queue: Mutex<VecDeque<Arc<Mutex<SockState>>>>,
-    afd_group: Mutex<Vec<Arc<Afd>>>,
+    lock: Mutex<()>,
+    cp: Arc<CompletionPort>,
+    active_poll_count: UnsafeCell<usize>,
+    update_queue: UnsafeCell<VecDeque<Arc<Mutex<SockState>>>>,
+    afd_group: AfdGroup,
 }
 
+// We have ensured thread safety by introducing lock manually.
+unsafe impl Sync for SelectorInner {}
+
 impl SelectorInner {
     pub fn new() -> io::Result<SelectorInner> {
-        CompletionPort::new(0).map(|cp| SelectorInner {
-            cp: cp,
-            active_poll_count: AtomicUsize::new(0),
-            update_queue: Mutex::new(VecDeque::new()),
-            afd_group: Mutex::new(Vec::new()),
+        CompletionPort::new(0).map(|cp| {
+            let cp = Arc::new(cp);
+            let cp_afd = Arc::clone(&cp);
+
+            SelectorInner {
+                lock: Mutex::new(()),
+                cp: cp,
+                active_poll_count: UnsafeCell::new(0),
+                update_queue: UnsafeCell::new(VecDeque::new()),
+                afd_group: AfdGroup::new(cp_afd),
+            }
         })
     }
 
     pub fn select(&self, events: &mut Events, timeout: Option<Duration>) -> io::Result<()> {
         events.clear();
 
-        self.update_sockets_events()?;
+        let deadline = match timeout {
+            Some(timeout) => Some(Instant::now() + timeout),
+            None => None,
+        };
+        let mut n = 0;
 
-        self.active_poll_count.fetch_add(1, Ordering::SeqCst);
+        loop {
+            if timeout.is_none() {
+                let len = self.select2(&mut events.statuses, &mut events.events, None)?;
+                if len == 0 {
+                    continue;
+                }
+                return Ok(());
+            } else {
+                if n >= events.statuses.len() {
+                    return Ok(());
+                }
+                let deadline = deadline.unwrap();
+                let now = Instant::now();
+                if now >= deadline {
+                    return Ok(());
```

Comment:
This behaviour is different compared to the Unix selectors. Here we return immediately if a timeout `Duration::from_secs(0)` would be passed, on unix we still do the system call. @carllerche you ok with the difference? I think it could prevent a use case were you don't want to block but still want to poll events and thus pass a zero duration (that is not none).

Reply:
We still need to do the syscall, just not block. We often call `poll` w/ duration 0 from Tokio because we want to fetch any pending IO events but don't want to block (we still have other work to do).

Reply:
PS, Nice catch @Thomasdezeeuw :+1:

Reply:
I think this is solved, right @PerfectLaugh?

Reply:
I didn't make the test for Some(Duration::from_secs(0)), so I don't know whether the behavior will be same as unix either.

Reply:
I've created a test in #1065.